main/refer.c

```diff
@@ -0,0 +1,527 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2023, Commend International
+ *
+ * Maximilian Fridrich <m.fridrich@commend.com>
+ *
+ * See http://www.asterisk.org for more information about
+ * the Asterisk project. Please do not directly contact
+ * any of the maintainers of this project for assistance;
+ * the project provides a web site, mailing lists and IRC
+ * channels for your use.
+ *
+ * This program is free software, distributed under the terms of
+ * the GNU General Public License Version 2. See the LICENSE file
+ * at the top of the source tree.
+ */
+
+/*! \file
+ *
+ * \brief Out-of-call refer support
+ *
+ * \author Maximilian Fridrich <m.fridrich@commend.com>
+ */
+
+/*** MODULEINFO
+	<support_level>core</support_level>
+ ***/
+
+#include "asterisk.h"
+
+#include "asterisk/_private.h"
+
+#include "asterisk/module.h"
+#include "asterisk/datastore.h"
+#include "asterisk/pbx.h"
+#include "asterisk/manager.h"
+#include "asterisk/strings.h"
+#include "asterisk/astobj2.h"
+#include "asterisk/vector.h"
+#include "asterisk/app.h"
+#include "asterisk/taskprocessor.h"
+#include "asterisk/refer.h"
+
+struct refer_data {
+	AST_DECLARE_STRING_FIELDS(
+		AST_STRING_FIELD(name);
+		AST_STRING_FIELD(value);
+	);
+};
+
+AST_LIST_HEAD_NOLOCK(outhead, refer_data);
+
+/*!
+ * \brief A refer.
+ */
+struct ast_refer {
+	AST_DECLARE_STRING_FIELDS(
+		/*! Where the refer is going */
+		AST_STRING_FIELD(to);
+		/*! Where we "say" the refer came from */
+		AST_STRING_FIELD(from);
+		/*! Where to refer to */
+		AST_STRING_FIELD(refer_to);
+		/*! An endpoint associated with this refer */
+		AST_STRING_FIELD(endpoint);
+		/*! The technology of the endpoint associated with this refer */
+		AST_STRING_FIELD(tech);
+	);
+	/* Whether to refer to Asterisk itself, if refer_to is an Asterisk endpoint. */
+	int to_self;
+	/*! Technology/dialplan specific variables associated with the refer */
+	struct ao2_container *vars;
+};
+
+/*! \brief Lock for \c refer_techs vector */
+static ast_rwlock_t refer_techs_lock;
+
+/*! \brief Vector of refer technologies */
+AST_VECTOR(, const struct ast_refer_tech *) refer_techs;
+
+static int refer_data_cmp_fn(void *obj, void *arg, int flags)
+{
+	const struct refer_data *one = obj, *two = arg;
+	if (one == NULL || one->name == NULL || two == NULL || two->name == NULL) {
+		return 0;
+	}
+	return !strcasecmp(one->name, two->name) ? CMP_MATCH | CMP_STOP : 0;
+}
+
+static void refer_data_destructor(void *obj)
+{
+	struct refer_data *data = obj;
+	ast_string_field_free_memory(data);
+}
+
+static void refer_destructor(void *obj)
+{
+	struct ast_refer *refer = obj;
+
+	ast_string_field_free_memory(refer);
+	ao2_cleanup(refer->vars);
+}
+
+struct ast_refer *ast_refer_alloc(void)
+{
+	struct ast_refer *refer;
+
+	if (!(refer = ao2_alloc(sizeof(*refer), refer_destructor))) {
```

Comment:
Does this need a lock? I don't see it used anywhere

Reply:
Great catch. I should lock it in `ast_refer_send` before calling `refer_send`.

I based my implementation off the messaging implementation. This is how it's done there, e.g. `message.c:432`.

Reply:
Messaging has interaction with the dialplan, this does not. This is a lower level  API. Can refer be modified by one thread while it is in use by another thread?

Reply:
It can not. The only `ast_refer_tech` that is implemented - the pjsip one - takes the `ast_refer` struct, creates its own `refer_data` out of it and then the `ast_refer` struct is no longer used.

Changing this to `ao2_alloc_options(sizeof(*refer), refer_destructor, AO2_ALLOC_OPT_LOCK_NOLOCK)`.